test/metabase/app_db/cluster_lock_test.clj

```diff
@@ -35,3 +39,42 @@
         (future (sut/with-cluster-lock ::test-lock (a/<!! fin-chan)))
         (future (Thread/sleep 500) (a/>!! fin-chan :done))
         (is (nil? (sut/with-cluster-lock ::test-lock (Thread/sleep 1))))))))
+
+(deftest concurrent-lock-creation-race-test
+  (testing "Two threads racing to create the same lock for the first time"
+    ;; h2 is not race-proof, but it is also not cross-process
+    (when (not= (mdb/db-type) :h2)
+      ;; Ensure that we are creating the lock for the first time
+      (t2/delete! :metabase_cluster_lock :lock_name (u/qualified-name ::race-test-lock))
+
+      (let [results   (atom [])
+            result!   (partial swap! results conj)
+            thread!   (fn [id ^CountDownLatch start-latch ^CountDownLatch end-latch thunk]
+                        (future
+                          (try
+                            (sut/with-cluster-lock ::race-test-lock
+                              (try
+                                (.countDown start-latch)
+                                (thunk)
+                                (result! [id :done])
+                                (catch Exception e
+                                  (result! [id :failed (:reason (ex-data e) :unknown)]))))
+                            (catch Exception e
+                              (result! [:a :locking-failed e]))
+                            (finally
+                              (.countDown end-latch)))))
+            a-started (CountDownLatch. 1)
```

Comment:
The other test uses channels, and they'd be fine here. I prefer latches though, since they express intent more clearly, and handle more complex scenarios more simply.